langchain/memory/vectorstore.py

```diff
@@ -0,0 +1,95 @@
+"""Class for a VectorStore-backed memory object."""
+
+from typing import Any, Dict, List, Optional, Union
+
+from pydantic import Field
+
+from langchain.memory.chat_memory import BaseMemory
+from langchain.memory.utils import get_prompt_input_key
+from langchain.schema import Document
+from langchain.vectorstores.base import VectorStore
+
+
+class VectorStoreMemory(BaseMemory):
+    """Class for a VectorStore-backed memory object."""
+
+    vectorstore: VectorStore = Field(exclude=True)
+    """Vector Database to connect to."""
+
+    k: int = 4
+    """Number of results to attempt to return from store."""
+
+    memory_key: str = "history"  #: :meta private:
+    """Key name to locate the memories in the result of load_memory_variables."""
+
+    input_key: Optional[str] = None
+    """Key name to index the inputs to load_memory_variables."""
+
+    search_type: str = "similarity"
+    """Method to use for searching."""
+
+    search_kwargs: Dict[str, Any] = Field(default_factory=dict)
+    """Extra search args."""
+
+    return_docs: bool = True
+    """Whether or not to return the result of querying the database directly."""
+
+    @property
+    def memory_variables(self) -> List[str]:
+        """The list of keys emitted from the load_memory_variables method."""
+        return [self.memory_key]
+
+    def _similarity_search(self, query: str, k: int) -> List[Document]:
+        """Search the database for similar documents."""
+        # TODO: Extract vectorstores with similarity_search_with_score classes
+        # to a separate mixin and check to attempt to guarantee orderedness
+        # of the output.
+        if self.search_type == "similarity":
+            docs = self.vectorstore.similarity_search(
+                query, k=self.k, **self.search_kwargs
+            )
+        elif self.search_type == "mmr":
+            docs = self.vectorstore.max_marginal_relevance_search(
+                query, k=self.k, **self.search_kwargs
+            )
+        else:
+            raise ValueError(f"search_type of {self.search_type} not allowed.")
+        return docs
```

Comment:
This doesn't have a guaranteed order but typically returns docs sorted by most-salient first.  If one were to use this as a chat memory object, that would be serialized as the most salient document being put towards the beginning of the conversation, which seems a bit off, but since we aren't using the _with_scores() method anyway, it seems like a. poor design decision to always return `docs[::-1]`